File: sdks/python/examples/05_multi_connection.py
#!/usr/bin/env python3
"""05 · Multi-connection and session adoption — two clients, one session.

The first client creates the session; the second joins it by id. Only the creator
ends it on disconnect, which is what makes adoption safe from a tab that may
close at any moment. A backend-created session is the same story.

    uv run python examples/05_multi_connection.py

Docs: https://docs.reactor.inc/concepts/sessions#multiple-connections-per-session
      https://docs.reactor.inc/concepts/sessions#adopting-an-existing-session
      https://docs.reactor.inc/concepts/sessions#who-owns-the-session
"""

from __future__ import annotations

import asyncio
import os
import sys
from collections.abc import Callable

import display

from reactor_sdk import DEFAULT_API_URL, Reactor

API_KEY = os.environ.get("REACTOR_API_KEY")
MODEL = os.environ.get("REACTOR_MODEL", "reactor/helios")
SECONDS = float(os.environ.get("REACTOR_SECONDS", "10"))
SHOW = os.environ.get("REACTOR_SHOW") == "1"

PROMPT = "a forest at dawn, sunbeams through the canopy"
OUTPUT_TRACK = "main_video"


def client(label: str) -> Reactor:
    reactor = Reactor(
        model_name=MODEL,
        api_key=API_KEY,
        api_url=os.environ.get("REACTOR_API_URL", DEFAULT_API_URL),
        local=os.environ.get("REACTOR_LOCAL") == "1",
    )

    @reactor.on_status
    def status(new: str) -> None:
        print(f"[{label}] status: {new}")

    @reactor.on_error
    def failed(error: Exception) -> None:
        print(f"[{label}] error: {error}")

    return reactor


def count_frames(reactor: Reactor, window, tile: int) -> Callable[[], int]:
    frames = 0

    @reactor.track(OUTPUT_TRACK).on_raw_frame
    def count(data: bytes, width: int, height: int, *_: object) -> None:
        nonlocal frames
        frames += 1
        window.submit(data, width, height, tile=tile)

    return lambda: frames


async def main() -> None:
    if not API_KEY and os.environ.get("REACTOR_LOCAL") != "1":
        sys.exit("set REACTOR_API_KEY — https://www.reactor.inc/account/api-keys")

    creator, joiner = client("creator"), client("joiner")
    # One tile per client: the same session, arriving twice.
    window = display.window(f"{MODEL} · creator | joiner", tiles=2, enabled=SHOW)

    async with creator, joiner:
        await creator.connect()
        print(f"session: {creator.session_id}")
        await creator.send_command("set_prompt", {"prompt": PROMPT})
        await creator.send_command("start", {})
        creator_frames = count_frames(creator, window, tile=0)

        # The id is the whole handoff — no second session, no coordination.
        await joiner.connect(session_id=creator.session_id)
        joiner_frames = count_frames(joiner, window, tile=1)

        await window.hold(SECONDS)
        print(f"frames creator: {creator_frames()}")
        print(f"frames joiner: {joiner_frames()}")

    print("both disconnected; the session ended with its creator")


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("interrupted")

File: sdks/cpp/examples/05_multi_connection.cpp
// 05 — Two clients on one session.
//
// The first creates the session and drives it. The second adopts it by id and
// only watches: same session, same generation, two independent connections.
//
// The teardown order is the lesson. The *creator* ends the session, so the
// watcher disconnects first — and both disconnect no matter what, because a
// creator that goes away without disconnecting leaves the session orphaned and
// the next run cannot start until it clears.
//
//   export REACTOR_API_KEY=...
//   ./05_multi_connection
//
// Docs: https://docs.reactor.inc/concepts/sessions

#include <atomic>
#include <chrono>
#include <cstdlib>
#include <iostream>
#include <reactor/reactor.hpp>
#include <thread>

#include "display.hpp"

namespace {

constexpr auto MODEL = "reactor/helios";

}  // namespace

int main() {
  const char* api_key = std::getenv("REACTOR_API_KEY");
  if (api_key == nullptr) {
    std::cerr << "set REACTOR_API_KEY\n";
    return 2;
  }

  reactor::Reactor creator{MODEL, reactor::ApiKey{api_key}};
  reactor::Reactor watcher{MODEL, reactor::ApiKey{api_key}};

  auto creator_errors = creator.on_error([](const reactor::ReactorError& error) {
    std::cerr << "creator error: " << error.what() << '\n';
  });
  auto watcher_errors = watcher.on_error([](const reactor::ReactorError& error) {
    std::cerr << "watcher error: " << error.what() << '\n';
  });

  int exit_code = 1;
  try {
    creator.connect().get();
    const auto session = creator.session_id().value();
    std::cout << "creator opened session " << session << '\n';

    creator.send_command("set_prompt", {{"prompt", "a paper boat on a wide river"}}).get();
    creator.send_command("start").get();

    // The same session, adopted rather than created. Everything else about this
    // client is ordinary.
    reactor::ConnectOptions adopt;
    adopt.session_id = session;
    watcher.connect(std::move(adopt)).get();
    std::cout << "watcher joined session " << watcher.session_id().value_or("?") << '\n';

    std::atomic<int> creator_frames{0};
    std::atomic<int> watcher_frames{0};
    examples::Display display{"05 — the watcher's view"};

    auto on_creator =
        creator.track("main_video").on_frame([&](const reactor::VideoFrame&) { ++creator_frames; });
    auto on_watcher = watcher.track("main_video").on_frame([&](const reactor::VideoFrame& frame) {
      ++watcher_frames;
      display.show(frame);
    });

    const auto deadline = std::chrono::steady_clock::now() + std::chrono::seconds(10);
    while (std::chrono::steady_clock::now() < deadline && display.pump()) {
      std::this_thread::sleep_for(std::chrono::milliseconds(20));
    }

    std::cout << "creator saw " << creator_frames.load() << " frames, watcher saw "
              << watcher_frames.load() << '\n';
    exit_code = (creator_frames.load() > 0 && watcher_frames.load() > 0) ? 0 : 1;
  } catch (const reactor::ReactorError& error) {
    std::cerr << "failed: " << error.what() << '\n';
  }

  // The watcher leaves first: disconnecting the creator ends the session for
  // everyone on it.
  try {
    watcher.disconnect().get();
  } catch (const reactor::ReactorError& error) {
    std::cerr << "watcher teardown: " << error.what() << '\n';
  }
  try {
    creator.disconnect().get();
  } catch (const reactor::ReactorError& error) {
    std::cerr << "creator teardown: " << error.what() << '\n';
  }
  return exit_code;
}
